src/service/crypto-verifier-impl.js

```diff
@@ -0,0 +1,198 @@
+
+const VERSION = 0x00;
+
+/**
+ * A service that provides cryptographic signature verification for use
+ * within AMP.
+ */
+class CryptoVerifier {
+  /**
+   * Verifies RSA signature corresponds to the data given a list of public keys.
+   * @param {!Uint8Array} the data that was signed.
+   * @param {!Uint8Array} the RSA signature.
+   * @param {{e: !Uint8Array, n:!Uint8Array, keyHash: !Uint8Array }} the RSA
+   *     public keys. If keyHash in each public key is null, this function will
+   *     compute it.
+   * @return {!Promise<!boolean>} whether the signature is valid for one of
+   *     the public keys.
+   */
+  verify(data, signature, rsaPubKeys) {
+    return hashRSAPubKeys(rsaPubKeys).then(hashes => {
+      for (let i = 0; i < rsaPubKeys.length; i++) {
+        const rsaPubKey = rsaPubKeys[i];
+        if (rsaPubKey['keyHash'] == null) {
+          rsaPubKey['keyHash'] = new Uint8Array(hashes[i], 0, 4);
+        }
+        const keyHash = rsaPubKey['keyHash'];
+        // The signature has the following format:
+        // 1-byte version + 4-byte key hash + raw RSA signature where
+        // the raw RSA signature is computed over (data || 1-byte version).
+        // As the key hash is unique, we verify against only one key.
+        if (signature.length > 5 && signature[0] == VERSION &&
+            hashesEqual(signature, keyHash)) {
+          const signedData = new Uint8Array(data.length + 1);
+          signedData.set(data);
+          signedData[data.length] = VERSION;
+          let jsonRsaKey = {
+            kty: 'RSA',
+            e: encodeByteArray(rsaPubKey['e'], true),
+            n: encodeByteArray(rsaPubKey['n'], true),
+            alg: 'RS256',
+            ext: true,
+          };
+          if (isWebkit) {
+            jsonRsaKey = asciiToUint8Array(JSON.stringify(jsonRsaKey));
+          }
+          return crossCrypto.importKey(
+            'jwk',
+            jsonRsaKey,
+            {
+              name: 'RSASSA-PKCS1-v1_5',
+              hash: {name: 'SHA-256'},
+            },
+            true,
+            ['verify'])
+
+            .then(publicKey => crossCrypto.verify({
+              name: 'RSASSA-PKCS1-v1_5',
+              hash: 'SHA-256',
+            }, publicKey, new Uint8Array(signature, 5), signedData))
+
+            .catch(err => {
+              dev.error(TAG_, 'Error while verifying:', err);
+              throw err;
+            });
+        };
+      };
+      return Promise.resolve(false);
+    });
+  }
+
+  /**
+   * Is this service actually available? For now, we use browser native
+   * crypto. So if that is not available, then this service is not available.
+   * @return {boolean}
+   */
+  isAvailable() {
+    return !!crossCrypto;
+  }
+}
+
+/**
+ * Appends 4-byte endian data's length to the data itself.
+ * @param {!Uint8Array} the data.
+ * @return {!Uint8Array} the appended 4-byte endian data's length together with
+ *     the data itself.
+ */
+function lenPrefix(data) {
```

Comment:
This kind of went above my "obvious" threshold and it is perfectly unit testable on the other hand.